From 2ce2476ed448711ab7e565bbdf3ade229e880542 Mon Sep 17 00:00:00 2001
From: Pankaj Dev <pankaj.dev@st.com>
Date: Thu, 2 Feb 2023 09:37:21 +0530
Subject: [PATCH 033/120] usb: dwc3: Call set_mode for the PHY

add support to call set_mode callback of all the PHYs
linked to the dwc3

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git e8c80b37787463488efa67015198e1d8fb9d0d35 usb: dwc3: Call set_mode for the PHY]
Signed-off-by: Pankaj Dev <pankaj.dev@st.com>
Change-Id: I52de538f8db20572cd26e15d473b660d34db2a84
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/303894
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/usb/dwc3/core.c | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 847fa1f82c3..a594e98fb6c 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -1021,6 +1021,8 @@ int dm_usb_gadget_handle_interrupts(struct udevice *dev)
 int dwc3_setup_phy(struct udevice *dev, struct phy_bulk *phys)
 {
 	int ret;
+	enum usb_dr_mode dr_mode;
+	enum phy_mode phymode;
 
 	ret = generic_phy_get_bulk(dev, phys);
 	if (ret)
@@ -1032,7 +1034,32 @@ int dwc3_setup_phy(struct udevice *dev, struct phy_bulk *phys)
 
 	ret = generic_phy_power_on_bulk(phys);
 	if (ret)
-		generic_phy_exit_bulk(phys);
+		goto err_power_on;
+
+	dr_mode = usb_get_dr_mode(dev_ofnode(dev));
+
+	switch(dr_mode)
+	{
+		case USB_DR_MODE_HOST:
+			phymode = PHY_MODE_USB_HOST;
+			break;
+		case USB_DR_MODE_PERIPHERAL:
+			phymode = PHY_MODE_USB_DEVICE;
+			break;
+		default:
+			goto err_mode;
+	}
+
+	ret = generic_phy_set_mode_bulk(phys, phymode, 0);
+	if (ret)
+		goto err_mode;
+
+	return ret;
+
+err_mode:
+	generic_phy_power_off_bulk(phys);
+err_power_on:
+	generic_phy_exit_bulk(phys);
 
 	return ret;
 }
-- 
2.47.2

